### PR TITLE
nbd: Add NBD (Network Block Device) support.

### PR DESCRIPTION
**Pull request by rwmjones**
Only the NBD "fixed newstyle" protocol is supported, since the
oldstyle protocol is now obsolete.

This is only a partial implementation of the protocol, enough to
access non-TLS servers and upload/download data only.  For other
non-supported items see the list at the top of lib/nbd.c.

Signed-off-by: Richard W.M. Jones <rjones@redhat.com>

### CODE REVIEW

File: lib/nbd.c [hunk 1]
```diff
@@ -0,0 +1,792 @@
...
+#include "urldata.h"
+#include "connect.h"
+#include "progress.h"
+#include "sendf.h"
+#include "multiif.h"
+#include "transfer.h"
+#include "escape.h"
+#include "nbd.h"
+/* The last #include file should be: */
+#include "memdebug.h"
```
Comment (bagder) [inline lib/nbd.c]: I recommend stealing the three last #include files from for example lib/url.c:
~~~c
/* The last 3 #include files should be in this order */
#include "curl_printf.h"
#include "curl_memory.h"
#include "memdebug.h"

File: lib/nbd.c [hunk 2]
```diff
@@ -0,0 +1,792 @@
...
+#include <in.h>
+#include <inet.h>
+#endif
+
+#if (defined(NETWARE) && defined(__NOVELL_LIBC__))
+#undef in_addr_t
+#define in_addr_t unsigned long
+#endif
+
+#include <curl/curl.h>
```
Comment (bagder) [inline lib/nbd.c]: Unnecessary, already done by `curl_setup.h`

File: lib/nbd.c [hunk 3]
```diff
@@ -0,0 +1,792 @@
...
+#include "nbd.h"
+/* The last #include file should be: */
+#include "memdebug.h"
+
+/* Maximum we will read from the socket, which is also the size of
+ * the internal recv_buf buffer that we allocate per handle.
+ */
+#define NBD_MAX_MESSAGE_SIZE 8192
+
+/* The maximum size of a request we will make to the server.
```
Comment (bagder) [inline lib/nbd.c]: Is this the maximum size of an upload? Or what is "a request" in this context?

File: lib/nbd.c [hunk 4]
```diff
@@ -0,0 +1,792 @@
...
+static CURLcode nbd_connect(struct connectdata *conn, bool *done)
+{
+  struct nbd_conn *nbdc = &conn->proto.nbdc;
+  CURLcode result;
+  (void)done;
+
+  /* Initialize the connection structure. */
+  memset(nbdc, 0, sizeof(*nbdc));
+  state(conn, NBD_EXPECTING_MAGIC);
+  nbdc->recv_buf = malloc(NBD_MAX_MESSAGE_SIZE);
```
Comment (bagder) [inline lib/nbd.c]: Can't the ndb code use the existing read buffer libcurl already allocates and use for all other protocols? I realize smb.c uses its own buffer, but that was never a good design choice there and we should work on removing that. Unless there's a particular need, there's already an allocated buffer for reading remote resources into.

File: lib/nbd.c [hunk 5]
```diff
@@ -0,0 +1,792 @@
...
+  struct handshake *handshake = NULL;
+  CURLcode result;
+
+  result = nbd_send_recv(conn, sizeof(struct handshake), (void **) &handshake);
+  if(result)
+    return result;
+  if(!handshake)
+    return CURLE_OK;
+
+  if(memcmp(handshake->nbdmagic, "NBDMAGIC", 8) != 0 ||
```
Comment (bagder) [inline lib/nbd.c]: What if it reads less than 8 bytes in the first read?

File: lib/nbd.c [hunk 6]
```diff
@@ -0,0 +1,792 @@
...
+  const bool uploading = conn->data->set.upload;
+  CURLcode result;
+
+  result = nbd_send_recv(conn, sizeof(struct export_data), (void **) &data);
+  if(result)
+    return result;
+  if(!data)
+    return CURLE_OK;
+
+  nbdc->size = be64toh(data->size);
```
Comment (bagder) [inline lib/nbd.c]: Is this (blindly) trusting that the remote server sends a valid size? What happens if it lies/is malicious?

File: lib/nbd.c [hunk 7]
```diff
@@ -0,0 +1,792 @@
...
+struct option {
+  char version[8];              /* "IHAVEOPT"  */
+  unsigned int option;          /* NBD_OPT_* */
+#define NBD_OPT_EXPORT_NAME 1
+  unsigned int len;             /* Length of option data. */
+  char data[1];                 /* Option data. */
+} PACK;
+
+struct export_data {
+  curl_off_t size;              /* Size of the export. */
```
Comment (bagder) [inline lib/nbd.c]: You use `curl_off_t` in a packed struct, which tells me you assume a fixed size here. That's not something `curl_off_t` guarantees, so if this code only works with a 64 bit type, you need to protect the entire code with an #ifdef for that. (smb.c does that)

**Comment by gvanem** [comment]
How about something for `version.c` too? Like:
```diff
--- a/version.c 2018-05-22 08:43:49
+++ b/version.c 2018-07-20 23:23:23
@@ -294,6 +294,9 @@
 #ifndef CURL_DISABLE_TFTP
   "tftp",
 #endif
+#ifndef CURL_DISABLE_NBD
+  "ndb",
+#endif

   NULL
 };
```

**Comment by danielgustafsson** [comment]
@rwmjones are you intending to work on this?

**Comment by bagder** [comment]
No comment since July. Closing.

**Pull request closed without merging**